File: apps/files/InspectionKeys.js
.pragma library

function normalized(event) {
    switch (event.key) {
    case Qt.Key_Space: return " ";
    case Qt.Key_Escape: return "Escape";
    case Qt.Key_Return: return "Return";
    case Qt.Key_Enter: return "Enter";
    default: return event.text;
    }
}

function press(event, controller, popup) {
    // The modal Quick Look popup blocks window Shortcuts, so Ctrl+O/Ctrl+I are forwarded from here
    // (navigation-history DESIGN.md §4.3); elsewhere Main.qml's Shortcuts handle them.
    if (popup && (event.modifiers & Qt.ControlModifier) && (event.key === Qt.Key_O || event.key === Qt.Key_I)) {
        if (event.key === Qt.Key_O)
            controller.navigateHistoryBack();
        else
            controller.navigateHistoryForward();
        event.accepted = true;
        return;
    }
    const key = normalized(event);
    if (popup && key !== " " && key !== "Escape" && key !== "j" && key !== "k")
        return;
    if (key === " " && event.isAutoRepeat) {
        event.accepted = true;
        return;
    }
    if (key.length > 0 && controller.handleKey(key))
        event.accepted = true;
}

function release(event) {
    if (event.key === Qt.Key_Space)
        event.accepted = true;
}

function overrideShortcut(event, popup, blockEscape) {
    // blockEscape: true while VISUAL mode is active (Escape there must exit VISUAL, not leave
    // fullscreen) — INSERT/SEARCH move focus onto their own text field, which claims Escape
    // itself, so they never need this parameter.
    if (event.key === Qt.Key_Space || (event.key === Qt.Key_Escape && (popup || blockEscape)))
        event.accepted = true;
}

File: apps/files/presentation/inspection_keys.cpp
#include "inspection_keys.h"

namespace {
QString normalized(int key, const QString& text) {
  switch (key) {
    case Qt::Key_Space:
      return QStringLiteral(" ");
    case Qt::Key_Escape:
      return QStringLiteral("Escape");
    case Qt::Key_Return:
      return QStringLiteral("Return");
    case Qt::Key_Enter:
      return QStringLiteral("Enter");
    default:
      return text;
  }
}
}  // namespace

// QML invokes this method through the engine-owned singleton instance.
// NOLINTNEXTLINE(readability-convert-member-functions-to-static)
bool InspectionKeys::press(int key, const QString& text, int modifiers, bool autoRepeat,
                           DirectoryController* controller, bool popup) const {
  if (popup && (modifiers & Qt::ControlModifier) != 0 && (key == Qt::Key_O || key == Qt::Key_I)) {
    if (controller != nullptr) {
      if (key == Qt::Key_O) {
        controller->navigateHistoryBack();
      } else {
        controller->navigateHistoryForward();
      }
    }
    return true;
  }
  const auto input = normalized(key, text);
  if (popup && input != " " && input != "Escape" && input != "j" && input != "k") {
    return false;
  }
  if (input == " " && autoRepeat) {
    return true;
  }
  return !input.isEmpty() && controller != nullptr && controller->handleKey(input);
}

// QML invokes this method through the engine-owned singleton instance.
// NOLINTNEXTLINE(readability-convert-member-functions-to-static)
bool InspectionKeys::release(int key) const { return key == Qt::Key_Space; }

// QML invokes this method through the engine-owned singleton instance.
// NOLINTNEXTLINE(readability-convert-member-functions-to-static)
bool InspectionKeys::overrideShortcut(int key, bool popup, bool blockEscape) const {
  return key == Qt::Key_Space || (key == Qt::Key_Escape && (popup || blockEscape));
}
